src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -204,73 +204,42 @@ public LayoutUpdateVisitor(List<Pathfinder.PathElement> path, LayoutRule args,
+                for (int i = 0; i < originalValue.length; i++) {
+                    originalParams.addRule(i, originalValue[i]);
                 }
+                changedView.setLayoutParams(originalParams);
             }
             mAlive = false;
         }
 
         @Override
         public void visit(View rootView) {
+            // this check is necessary - if the layout change is invalid, accumulate will send an error message
+            // to the Web UI; before Web UI removes such change, this visit may get called by Android again and
+            // thus send another error message to Web UI which leads to lots of weird problems
             if (mAlive) {
                 getPathfinder().findTargetsInRoot(rootView, getPath(), this);
             }
         }
 
+        // layout changes are performed on the children of found according to the LayoutRule
         @Override
         public void accumulate(View found) {
-            final int newVerb = mArgs.verb;
-            final int newAnchorId = mArgs.anchor;
-            final RelativeLayout.LayoutParams currentParams = (RelativeLayout.LayoutParams)found.getLayoutParams();
-            final int[] currentRules = currentParams.getRules();
-
-            if (currentRules[newVerb] == newAnchorId) {
-                return;
-            }
-
-            if (mOriginalValues.containsKey(found)) {
-                ; // Cache exactly one
-            } else {
-                LayoutRule originalValue = new LayoutRule(newVerb, currentRules[newVerb]);
-                mOriginalValues.put(found, originalValue);
-            }
-
             try {
-                setLayout(found, newVerb, newAnchorId);
+                setLayout(found);
             } catch (CantVisitException e) {
                 cleanup();
                 mOnEditErrorListener.onError(e);
             }
         }
 
-        private void setLayout(View target, int verb, int anchorId) throws CantVisitException {
-            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
-            params.addRule(verb, anchorId);
-
-            final Set<Integer> rules;
-            if (mHorizontalRules.contains(verb)) {
-                rules = mHorizontalRules;
-            } else if (mVerticalRules.contains(verb)) {
-                rules = mVerticalRules;
-            } else {
-                rules = null;
-            }
-
-            if (rules != null && !verifyLayout(target, rules)) {
-                throw new CantVisitException("Circular dependency detected!", "circular_dependency", mName);
-            }
-
-            target.setLayoutParams(params);
-        }
-
-        private boolean verifyLayout(View target, Set<Integer> rules) {
-            ViewGroup parent = (ViewGroup) target.getParent();
+        private void setLayout(View target) throws CantVisitException {
```

Comment:
This could probably be inlined, since it is only called once and takes no arguments.